### ISSUE TITLE
ScipyOptimizerInterface Support

### ISSUE THREAD

**Issue by cshenton** [blei-lab/edward#763]
Some VI problems I'm working have a simple enough latent variable space that I suspect L-BFGS-B will work better than SGD algorithms.

However, the scipy interface's API is different to basic optimizers (it requires the loss to define), so I can't just pass it in as `VariationalInference` is currently defined.
```python
optimizer = ScipyOptimizerInterface(loss, options={'maxiter': 100})
```
In `VariationalInference.initialize()`, by the time the optimizer arg is being parsed, the loss has been defined. A simple approach for now could be to just support L-BFGS (the default optimizer in this interface) without any extra options.

```python
...
elif optimizer == 'L-BFGS-B':
    optimizer = ScipyOptimizerInterface(self.loss)
...
```
And then longer term, think about general `ExternalOptimizerInterface` support (which all have the challenge of requiring a loss to instantiate).

Would it be alright if I opened a PR to this effect?

**Comment by cshenton** [comment]
This would also require changing initialize to not try and grab the `apply_gradients` op (since external interfaces only have minimize), and adding a run method that just calls minimize if a scipy optimizer is detected (else calls `super().run()`)

**Comment by dustinvtran** [comment]
> Would it be alright if I opened a PR to this effect?

Absolutely. You'll probably have to change the logic a bit so that `optimizer` doesn't try to call `apply_gradients` if it's a `SciPyOptimizerInterface` object.

I guess `ScipyOptimizerInterface` suffers from the same unfortunate problem as our PrettyTensor optimizers in that you must pass a surrogate loss function from which it'll take gradients/etc. through rather than being able to pass in the gradients directly. It's worth noting in the docs when you use SciPyOptimizerInterface that this happens, as the autodiffed loss is not always the gradient we'd like to update with.

**Comment by cshenton** [comment]
So it's currently possible to manually grab the loss and use an external optimizer interface.
```python
import edward as ed
import numpy as np
import tensorflow as tf

from edward.models import Normal

MU = 5.0
SD = 1.0
N = 100000
y_obs = np.random.normal(MU, SD, N)

mu = Normal(0.0, 5.0)
y = Normal(mu, SD, sample_shape=N)

inference = ed.MAP([mu], {y: y_obs})
inference.initialize()
optimizer = tf.contrib.opt.ScipyOptimizerInterface(
    inference.loss,
    options={'eps': 1e-12, 'ftol': 1e-10})
tf.global_variables_initializer().run()
optimizer.minimize()

print('Predicted mu:', ed.get_session().run(list(inference.latent_vars.values())))
```

Having done some tests with this, it only seems to be worth it for MAP estimation, and only using the L-BFGS-B algorithm. Otherwise you're better off sticking to SGD. I bring this up because the `ScipyOptimizerInterface` only has a `minimize` method, so it doesn't fit in at all with the init-update approach of the existing estimators, so looks like it would be very awkward to integrate.

I think there's two options:
- I put the above snippet together as a tutorial, and people with specific use cases that benefit from this can use that.
- I make the required changes to initialize, and also add a run method to `VariationalInference` that first checks if a scipy optimizer is being used (and runs init-minimize) else defers to the parent class's run method (which runs the usual init-update-etc).

**Comment by dustinvtran** [comment]
> ScipyOptimizerInterface only has a minimize method, so it doesn't fit in at all with the init-update approach of the existing estimators, so looks like it would be very awkward to integrate.

That makes sense.

> I put the above snippet together as a tutorial, and people with specific use cases that benefit from this can use that.

IMO this is the better option. There's too many nuances in the alternative (surrogate loss whose autodiff are the desired gradients; can only call `run()` but manually do updates) for users to understand when and how to make it applicable. I suspect for a more native integration, we'll just have to wait until a LBFGS-B optimizer is written in TF.

**Comment by cshenton** [comment]
Glad you agree. I could see how it could be done, but I was also fairly certain it would really foul up the tractable class hierarchy in `ed.inference`. 

I'll put together that tutorial, but it might be a little slow going, since I need to brush up on my latex.
